utils/tail/tail.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package tail implements "tail -F" functionality following rotated logs
+package tail
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/golang/glog"
+	"golang.org/x/exp/inotify"
+)
+
+type Tail struct {
+	reader   *bufio.Reader
+	filename string
+	file     *os.File
+	stop     chan bool
+	watcher  *inotify.Watcher
+}
+
+const retryReopenInterval = time.Second
+const maxReopenAttempts = 3
+
+// NewTail starts opens the given file and watches it for deletion/rotation
+func NewTail(filename string) (t *Tail, err error) {
+	t = &Tail{
+		filename: filename,
+	}
+	t.file, err = os.Open(t.filename)
+	if err != nil {
+		glog.Errorf("Open failed on %s: %v", t.filename, err)
+		return nil, err
+	}
+	// TODO: not interested in old events?
+	//t.file.Seek(0, os.SEEK_END)
+	t.reader = bufio.NewReader(t.file)
+	t.stop = make(chan bool)
+	t.watcher, err = inotify.NewWatcher()
+	if err != nil {
+		glog.Errorf("Inotify init failed on %s: %v", t.filename, err)
+		return
+	}
+	go t.watchFile()
+	return t, nil
+}
+
+// Read implements the io.Reader interface for Tail
+func (t *Tail) Read(p []byte) (n int, err error) {
+	n, err = t.reader.Read(p)
+	if err != nil {
+		return 0, io.EOF
+	}
+	return n, nil
+}
+
+var _ io.Reader = &Tail{}
+
+// Close stops watching and closes the file
+func (t *Tail) Close() {
+	close(t.stop)
+}
+
+func isEvent(event *inotify.Event, flag uint32) bool {
+	if event.Mask&flag == flag {
+		return true
+	}
+	return false
+}
+
+func (t *Tail) fileChanged() error {
+	for {
+		select {
+		case event := <-t.watcher.Event:
+			// We don't get IN_DELETE because we are holding the file open
+			if isEvent(event, inotify.IN_ATTRIB) || isEvent(event, inotify.IN_MOVE_SELF) {
+				return nil
+			}
+		case <-t.stop:
+			return fmt.Errorf("Watch was cancelled")
+		}
+	}
+}
+
+func (t *Tail) attemptReopen() (err error) {
+	for attempt := 1; attempt <= maxReopenAttempts; attempt++ {
+		select {
+		case <-time.After(retryReopenInterval):
+		case <-t.stop:
+			return fmt.Errorf("Watch was cancelled")
+		}
+		glog.V(4).Infof("Reopening %s (attempt %d of %d)", t.filename, attempt, maxReopenAttempts)
+		t.file, err = os.Open(t.filename)
+		if err == nil {
+			return nil
+		}
+	}
+	return err
+}
+
+func (t *Tail) watchFile() {
+	defer t.file.Close()
```

Comment:
This will close whatever `t.file` was when `WatchFile` _started_, not when it _finished_. See [defer](https://golang.org/doc/effective_go.html#defer).

I suggest flipping the inner loop segments (open the file then watch then close -- rather than watch, close, open), and encapsulating the loop body in a function, and doing `defer t.file.Close()` in that function.